### PR TITLE
Worked on vector-related performance

### PR DESCRIPTION
**Pull request by Kixiron**
* Pre-allocated vectors wherever possible in order to reduce the total amount of allocations preformed

### CODE REVIEW

File: src/db/add_package.rs [hunk 1]
```diff
@@ -212,13 +212,15 @@ fn initialize_package_in_database(conn: &Connection, pkg: &MetadataPackage) -> R
 
 /// Convert dependencies into Vec<(String, String, String)>
 fn convert_dependencies(pkg: &MetadataPackage) -> Vec<(String, String, String)> {
-    let mut dependencies: Vec<(String, String, String)> = Vec::new();
+    let mut dependencies: Vec<(String, String, String)> = Vec::with_capacity(pkg.dependencies.len());
+
```
Comment (jyn514) [inline src/db/add_package.rs:216]: If you rewrote this as a `map`, `collect` would preallocate automatically.

File: src/db/add_package.rs [hunk 2]
```diff
@@ -443,6 +445,8 @@ fn get_owners(pkg: &MetadataPackage) -> Result<Vec<CrateOwner>> {
     if let Some(owners) = json.as_object()
         .and_then(|j| j.get("users"))
         .and_then(|j| j.as_array()) {
+            result.reserve(owners.len());
```
Comment (jyn514) [inline src/db/add_package.rs:448]: This indentation doesn't match the for loop below.

Comment (Kixiron) [reply src/db/add_package.rs:448]: Well, the indentation of that whole block is off, so whatever's fine

**Comment by Kixiron** [comment]
Changed the vast majority to maps, some really didn't lend to it

File: src/web/source.rs [hunk 3]
```diff
@@ -41,24 +41,24 @@ impl ToJson for FileList {
...
-                          },
-                          true.to_json());
-
-            file_vec.push(file_m.to_json());
-        }
-
-        m.insert("files".to_string(), file_vec.to_json());
+        let files = self.files.iter()
+            .map(|file| {
+                let mut file_map: BTreeMap<String, Json> = BTreeMap::new();
```
Comment (jyn514) [inline src/web/source.rs:46]: Can you do the same trick from `temp_json` above where you use the same temporary allocation for each file?

**Comment by jyn514** [comment]
Thanks so much for working on this! Little things like this are easy to forget about but they make a big difference :)

File: src/db/file.rs [hunk 4]
```diff
@@ -276,47 +277,56 @@ fn detect_mime(file_path: &Path) -> Result<&'static str> {
...
+            let json = temp_json.to_json();
+            temp_json.clear();
 
-    for file in file_list {
-        let mut v: Vec<String> = Vec::new();
-        v.push(file.0.clone());
-        v.push(file.1.into_os_string().into_string().unwrap());
-        file_list_json.push(v.to_json());
-    }
+            json
```
[the remark this answers is not in the document]
Comment (Kixiron) [reply src/db/file.rs:291]: Sorry, I just couldn't figure out what it really was based off of context (In reference to naming)

File: src/docbuilder/crates.rs [hunk 5]
```diff
@@ -14,7 +14,7 @@ fn crates_from_file<F>(path: &PathBuf, func: &mut F) -> Result<()>
     let reader = fs::File::open(path).map(|f| BufReader::new(f))?;
 
     let mut name = String::new();
-    let mut versions = Vec::new();
+    let mut versions = Vec::new(); // TODO: Find a way to pre-allocate the versions vector
```
[the remark this answers is not in the document]
Comment (Kixiron) [reply src/docbuilder/crates.rs:17]: Unfortunately I couldn't make it a map, although it's probably possible, because of the semantics between continuing and propagating errors

[the remark this answers is not in the document]
Comment (jyn514) [reply src/docbuilder/crates.rs:17]: Ok, just removing the comment is fine then.

File: src/web/source.rs (hunk 3 above)
Comment (Kixiron) [reply src/web/source.rs:46]: I'm not sure it'll have as much of an effect due to how BTrees work, but I'll certainly do it

**Comment by jyn514** [comment]
I would like to add tests for these before merging. See e.g. https://github.com/rust-lang/docs.rs/blob/master/src/web/mod.rs#L617 for examples of how to write tests, and you can visit the pages on docs.rs to see what the expected behavior should be.

File: src/web/source.rs [hunk 6]
```diff
@@ -101,48 +103,54 @@ impl FileList {
...
                         }
                     }
-                };
-            }
+
+                    None
+                })
+                .collect::<Vec<_>>();
+
+            file_list.dedup();
```
Comment (jyn514) [inline src/web/source.rs:150]: This does not preserve the same behavior. `dedup` only works if the array is already sorted, and these files may not be. Also, the `// avoid duplicates` comment should be here instead of above.

Comment (Kixiron) [reply src/web/source.rs:150]: My bad, that's a rather counter-intuitive implementation. Would you prefer me to sort or to revert to the previous implementation?

Comment (jyn514) [reply src/web/source.rs:150]: Let's revert, the PR is already big enough.

File: src/docbuilder/limits.rs [hunk 7]
```diff
@@ -151,40 +152,41 @@ mod test {
...
                 ..Limits::default()
             };
-            db.conn().query("INSERT INTO sandbox_overrides (crate_name, max_memory_bytes, timeout_seconds, max_targets)
-                                    VALUES ($1, $2, $3, $4)",
-                                    &[&krate, &(limits.memory as i64), &(limits.timeout.as_secs() as i32), &(limits.targets as i32)])?;
+            db.conn().query(
+                "INSERT INTO sandbox_overrides (crate_name, max_memory_bytes, timeout_seconds, max_targets)
+                 VALUES ($1, $2, $3, $4)",
+                &[&krate, &(limits.memory as i64), &(limits.timeout.as_secs() as i32), &(limits.targets as i32)]
+            )?;
```
Comment (jyn514) [inline src/docbuilder/limits.rs:159]: Please undo the formatting changes in this file, they're not related to the PR.

File: src/utils/github_updater.rs [hunk 8]
```diff
@@ -186,7 +186,7 @@ mod test {
         assert!(fields.is_ok());
 
         let fields = fields.unwrap();
-        assert!(fields.description != "".to_string());
+        assert!(fields.description != "");
```
Comment (jyn514) [inline src/utils/github_updater.rs:189]: Ditto, please revert formatting changes. These are good changes, if you put them in a separate PR they wouldn't be blocked by writing tests for the `map`s.

File: src/web/builds.rs [hunk 9]
```diff
@@ -69,46 +69,49 @@ pub fn build_list_handler(req: &mut Request) -> IronResult<Response> {
...
+                builds.rustc_version,
+                builds.cratesfyi_version,
+                builds.build_status,
+                builds.build_time,
+                builds.output
+         FROM builds
+         INNER JOIN releases ON releases.id = builds.rid
+         INNER JOIN crates ON releases.crate_id = crates.id
+         WHERE crates.name = $1 AND releases.version = $2
+         ORDER BY id DESC",
```
Comment (jyn514) [inline src/web/builds.rs:88]: Please revert formatting changes.

File: src/web/crate_details.rs [hunk 10]
```diff
@@ -115,34 +115,34 @@ impl CrateDetails {
...
+                crates.github_forks,
+                crates.github_issues,
+                releases.is_library,
+                releases.doc_targets,
+                releases.license,
+                releases.documentation_url,
+                releases.default_target
+         FROM releases
+         INNER JOIN crates ON releases.crate_id = crates.id
+         WHERE crates.name = $1 AND releases.version = $2;";
```
Comment (jyn514) [inline src/web/crate_details.rs:145]: Please revert formatting changes.

File: src/web/pool.rs [hunk 11]
```diff
@@ -40,7 +40,7 @@ impl BeforeMiddleware for Pool {
 }
 
 pub(crate) enum PoolConnection {
-    R2D2(r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>),
+    R2D2(Box<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>>),
```
Comment (jyn514) [inline src/web/pool.rs:43]: Why was this change made?

File: src/web/releases.rs [hunk 12]
```diff
@@ -602,26 +605,30 @@ pub fn activity_handler(req: &mut Request) -> IronResult<Response> {
...
-    for krate in &conn
+    let query = conn
         .query(
             "SELECT name, version, priority
-                              FROM queue
-                              WHERE attempt < 5
-                              ORDER BY priority ASC, attempt ASC, id ASC",
+                          FROM queue
+                          WHERE attempt < 5
+                          ORDER BY priority ASC, attempt ASC, id ASC",
```
Comment (jyn514) [inline src/web/releases.rs:613]: Please revert formatting changes.

File: src/web/rustdoc.rs [hunk 13]
```diff
@@ -446,21 +446,26 @@ impl Handler for SharedResourceHandler {
...
         assert_success(path, web)?;
         let data = web.get(path).send()?.text()?;
         let dom = kuchiki::parse_html().one(data);
-        for elems in dom.select("form ul li a.warn") {
-            for elem in elems {
-                let warning = elem.as_node().as_element().unwrap();
+
+        if let Ok(mut elements) = dom.select("form ul li a.warn") {
+            if let Some(element) = elements.next() {
+                let warning = element.as_node().as_element().unwrap();
```
Comment (jyn514) [inline src/web/rustdoc.rs:458]: This is a good change but it should go in a separate PR.

File: src/web/sitemap.rs [hunk 14]
```diff
@@ -8,21 +8,27 @@ use time;
...
-                                   crates.name,
-                                   releases.release_time
-                            FROM crates
-                            INNER JOIN releases ON releases.crate_id = crates.id
-                            WHERE rustdoc_status = true",
+                    crates.name,
+                    releases.release_time
+             FROM crates
+             INNER JOIN releases ON releases.crate_id = crates.id
+             WHERE rustdoc_status = true",
```
Comment (jyn514) [inline src/web/sitemap.rs:18]: Please revert formatting changes.

File: src/web/pool.rs (hunk 11 above)
Comment (Kixiron) [reply src/web/pool.rs:43]: It’s a size-difference thing, that variant is vastly larger than the other

Comment (jyn514) [reply src/web/pool.rs:43]: The other is only used in test code so I don't think we have to worry about a size difference.

**Comment by Kixiron** [comment]
Git hates me and closed on its own

**Comment by jyn514** [comment]
Status: this needs to be rebased and needs tests to be written.

File: src/web/sitemap.rs (hunk 14 above)
Comment (jyn514) [reply src/web/sitemap.rs:18]: I was being a little picky here, this is fine to keep.

File: src/web/releases.rs [hunk 15]
```diff
@@ -605,26 +608,30 @@ pub fn activity_handler(req: &mut Request) -> IronResult<Response> {
...
```
Comment (jyn514) [inline src/web/releases.rs:616]: If we're going to fix the formatting, let's fix it properly :laughing: 
```suggestion
             "SELECT name, version, priority 
              FROM queue
              WHERE attempt < 5
              ORDER BY priority ASC, attempt ASC, id ASC",
```

File: src/web/crate_details.rs (hunk 10 above)
Comment (jyn514) [reply src/web/crate_details.rs:145]: This one I would actually like you to revert, I think it looked better before.

File: src/web/source.rs (hunk 6 above)
Comment (jyn514) [reply src/web/source.rs:165]: @Kixiron can you revert this change please?

**Comment by Kixiron** [comment]
@jyn514 Fixed

**Comment by jyn514** [comment]
:tada: :tada:
